### PR TITLE
[tt-train] Bitwise-identical checkpointing

### LINKED ISSUE #47949: [tt-train] Ensure checkpointed training runs resume with bitwise-identical results
Currently checkpoints contain:

- model parameters
- optimizer state
- required configs

We still miss:

- RNG state
- dataloader state
- ?

Implement both and verify bitwise-identical checkpoint resume behavior. Add accompanying tests.

### PR DESCRIPTION
**Pull request by imichalakTT**
Closes https://github.com/tenstorrent/tt-metal/issues/47949

### Summary

Implemented bitwise-identical checkpointing by:
- checkpointing C++ autograd RNG state via straightforward capture/restore helpers
- checkpointing dataloader via get/set_state_dict and making dataset shuffling depend on local derived RNG from seed, epoch, and position, instead of a global RNG state

Added related tests.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:imichalak/train-byte-identical-checkpointing)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:imichalak/train-byte-identical-checkpointing)

### CODE REVIEW

**Review by abustamanteTT — APPROVED** [review]

File: tt-train/sources/ttml/autograd/auto_context.cpp [hunk 1]
```diff
@@ -19,6 +20,17 @@ void AutoContext::set_generator(const std::mt19937& generator) {
...
+std::string AutoContext::get_generator_state() const {
+    std::ostringstream oss;
+    oss << m_generator;  // mt19937's stream operator emits its full internal state (624 words + position)
+    return oss.str();
+}
+
+void AutoContext::set_generator_state(const std::string& state) {
+    std::istringstream iss(state);
+    iss >> m_generator;
+}
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/autograd/auto_context.cpp:35]: Agree

**Review by philei-tt — APPROVED** [review]
Looks good, thank you. Please check comment from copilot
